examples/advanced/quantum_volume.ipynb

```diff
@@ -50,38 +54,68 @@
     "# This cell contains the business logic that actually runs the quantum volume algorithm with\n",
     "# parameters specified in the previous cell.\n",
     "from examples.advanced import quantum_volume\n",
+    "from collections import defaultdict\n",
     "import numpy as np\n",
     "\n",
-    "# Reset the probabilities\n",
     "for sampler in samplers:\n",
-    "    sampler['probabilities'] = [0] * len(depths)\n",
+    "    sampler['probabilities'] = defaultdict(int)\n",
+    "    sampler['compiled-probabilities'] = defaultdict(int)\n",
+    "    sampler['optimized-probabilities'] = defaultdict(int)\n",
     "\n",
     "for depth in depths:\n",
     "    num_qubits = depth  # Square matrix.\n",
     "    print(f\"Running simulation with {num_qubits} qubits and a depth of {depth}\")\n",
     "    for i in range(num_repetitions):\n",
     "        print(f\"    Repetition {i + 1} of {num_repetitions}\")\n",
-    "        # Generate a model circuit.\n",
+    "        # Generate a model circuit and compute its heavy set.\n",
     "        model_circuit = quantum_volume.generate_model_circuit(\n",
     "            num_qubits, depth, random_state=np.random.RandomState())\n",
-    "        # Compute its heavy set.\n",
     "        heavy_set = quantum_volume.compute_heavy_set(model_circuit)\n",
     "        print(f\"        Heavy Set: {heavy_set}\")\n",
     "\n",
-    "        # Run the given samplers over the model circuit.\n",
+    "        # Compile and optimize the model circuit.\n",
+    "        [compiled_circuit,\n",
+    "         mapping] = quantum_volume.compile_circuit(model_circuit, device=device)\n",
+    "        optimized_circuit = optimize(compiled_circuit)\n",
+    "\n",
+    "        # Run the given samplers over the model, compiled, and optimized circuits.\n",
     "        for sampler in samplers:\n",
     "            probability = quantum_volume.sample_heavy_set(\n",
     "                model_circuit, heavy_set, sampler=sampler['sampler'])\n",
-    "            sampler['probabilities'][depth - depths[-1]] += probability\n",
+    "            sampler['probabilities'][depth] += probability\n",
     "            print(f\"        {sampler['label']} HOG probability: {probability}\")\n",
     "\n",
-    "        # TODO: The next steps are to compile the model circuit and run it on various devices.\n",
+    "            compiled_probability = quantum_volume.sample_heavy_set(\n",
```

Comment:
outside the scope of this PR:

If you refactor the code into a problem generation task and data collection task (as suggested in another comment), you can probably simplify the notebook and just call those functions. Obviously this is a balance between making the notebook instructive (i.e. not a complete black box) and concise

Reply:
Yea I struggled with how much to reproduce in here to make it instructional. I agree that it could be cleaned up a bit - as above, I'd prefer to do it in the next PR if that's OK.